supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala

```diff
@@ -0,0 +1,146 @@
+    }.toMap
+
+    val filesWithValidPaths = filesAndValidatedPaths.collect {
+      case (referenceFile, validPath: ValidFullGcsPath) => (referenceFile, validPath)
+    }
+    val filesWithInvalidPaths = filesAndValidatedPaths.collect {
+      case (referenceFile, invalidPath: InvalidFullGcsPath) => (referenceFile, invalidPath)
+    }
+
+    if (filesWithInvalidPaths.nonEmpty) {
+      IO.raiseError(new RuntimeException(s"Some of the paths in manifest file are not valid GCS paths: " +
+        s"${filesWithInvalidPaths.keySet.map(_.path).toList.mkString(", ")}")
+      )
+    } else {
+      IO {
+        val gcsBatch = gcsClient.batch()
+        val filesAndBlobResults = filesWithValidPaths.map {
+          case (referenceFile, ValidFullGcsPath(bucket, path)) =>
+            val blobGetResult = gcsBatch.get(BlobId.of(bucket, path.substring(1)), BlobGetOption.fields(BlobField.CRC32C))
+            (referenceFile, blobGetResult)
+        }
+        gcsBatch.submit()
+
+        filesAndBlobResults.map {
+          case (referenceFile, blobGetResult) =>
+            val crc32cFromManifest = BaseEncoding.base64.encode(
+              // drop 4 leading bytes from Long crc32c value
+              // https://stackoverflow.com/a/25111119/1794750
+              util.Arrays.copyOfRange(Longs.toByteArray(referenceFile.crc32c), 4, 8)
+            )
+
+            (referenceFile, crc32cFromManifest === blobGetResult.get().getCrc32c)
+        }
+      }
+    }
+  }
+
+  private def manifestFileToMapOfValidReferences(manifestFile: ManifestFile): IO[Map[String, PipelinesApiReferenceFilesDisk]] = {
+    val refDisk = PipelinesApiReferenceFilesDisk(manifestFile.imageIdentifier, manifestFile.diskSizeGb)
+    val allReferenceFilesFromManifestMap = manifestFile.files.map(refFile => (refFile, refDisk)).toMap
+    val filesWithValidatedCrc32csIo = bulkValidateCrc32cs(allReferenceFilesFromManifestMap.keySet)
+
+    val validReferenceFilesFromManifestMapIo = filesWithValidatedCrc32csIo map {
+      filesWithValidatedCrc32c => {
+        allReferenceFilesFromManifestMap.filterKeys(key => filesWithValidatedCrc32c.getOrElse(key, false))
+      }
+    }
+
+    validReferenceFilesFromManifestMapIo map { validReferenceFilesFromManifestMap =>
+      val invalidReferenceFiles = allReferenceFilesFromManifestMap.keySet -- validReferenceFilesFromManifestMap.keySet
+      if (invalidReferenceFiles.nonEmpty) {
+        logger.warn(s"The following files listed in references manifest have checksum mismatch with actual files in GCS: ${invalidReferenceFiles.mkString(",")}")
+      }
+      validReferenceFilesFromManifestMap map {
+        case (refFile, disk) => (refFile.path, disk)
+      }
+    }
+  }
+}
+
+object PipelinesApiReferenceFilesMapping {
+  case class ReferenceFile(path: String, crc32c: Long)
+  case class ManifestFile(imageIdentifier: String, diskSizeGb: Int, files: List[ReferenceFile])
```

Comment:
These should all be `final case class`es throughout

Reply:
After replacing constructor initialization with case class and `generateReferenceFilesMapping` method in the companion object per suggestion above, I finally had to extract logic into a trait to make it testable, but now, when these 2 classes are inside a trait, it's impossible to make them final - it results in compilation error `Error:(36, 20) The outer reference in this type test cannot be checked at run time`. So we either leave them as is, or there should be another way to test the  logic.

Reply:
Did you try putting these into a static `object` instead of a trait? I'm not sure what happens when you try to access trait-defined classes from different subtypes of that trait... it seems like maybe you just cannot!

Reply:
Currently they are inside the trait, but it's impossible to make them private.